Add pytest fixtures support with Pylance

luabud opened the pull request:
This feature is now on the latest Pylance stable version: https://devblogs.microsoft.com/python/python-in-visual-studio-code-february-2023-release/#preview-improved-intellisense-support-for-pytest-with-pylance

luabud commented:
cc @rchiodo @heejaechang @debonte @bschnurr

gregvanl commented on docs/python/testing.md:
```diff
@@ -313,6 +313,21 @@ Below are all the supported commands for testing with the Python extension in VS
+
+As you're typing the parameters for your test function, Pylance will offer you a list of [auto completions](/docs/python/editing.md#autocomplete-and-intellisense) that includes known parameters from arguments defined by `@pytest.mark.parametrize`, as well as existing pytest fixtures defined in your tests file or in `confest.py`. [Code navigation](/docs/python/editing.md#navigation) features such  ...
+
+![Auto completion suggestion when passing a parameter to a test function in a Python test file. The suggestion is a fixture defined in conftest.py.](/docs/python/images/testing/pytest-fixture-autocomplete.png)
```
You can just use `images/testing/pytest-fixture-autocomplete.png` for the image paths (here and below).

gregvanl commented on docs/python/testing.md:
```diff
@@ -313,6 +313,21 @@ Below are all the supported commands for testing with the Python extension in VS
 |  **Testing: Focus on Test Explorer View** | Open the Test Explorer view. Similar to **Testing: Focus on Python View** on versions prior to 2021.9.
 |  **Test: Stop Refreshing Tests** | Cancel test discovery. |
 
+## IntelliSense for pytest fixtures and parameterization
```
This H2 text will likely result in word wrap in the right navigation "In this topic" list.
Could it be shortened to just "IntelliSense for pytest"?

rchiodo commented:
The image for the parameter types is not actually using the feature because the type annotation is already present. Granted the type annotation gives a better result, but it should show a hover tip like:

````
(parameter) method : Literal['fetch', ....
```

brettcannon commented on docs/python/testing.md:
```diff
@@ -313,6 +313,21 @@ Below are all the supported commands for testing with the Python extension in VS
+
+![Pylance type inference based on the types of arguments passed to the paramaterization decorator.](/docs/python/images/testing/pytest-inferred-parametrized-argument.png)
+
+Pylance also offers [code actions](/docs/editor/refactoring.md#code-actions--quick-fixes-and-refactorings) to add type annotation to test functions that has  fixture parameters.  Inlay hints for inferred fixture parameter types can also be enabled by setting `python.analysis.inlayHints.pytestParameters` to `true` in your User settings.
```
```suggestion
Pylance also offers [code actions](/docs/editor/refactoring.md#code-actions--quick-fixes-and-refactorings) to add type annotations to test functions that have fixture parameters.  Inlay hints for inferred fixture parameter types can also be enabled by setting `python.analysis.inlayHints.pytestParameters` to `true` in your User settings.
```

brettcannon commented on docs/python/settings-reference.md:
```diff
@@ -66,6 +66,7 @@ The language server settings apply when `python.languageServer` is `Pylance` or
 | importFormat | `absolute`| Defines the default format when auto importing modules. Accepted values are `absolute` or `relative`. |
 | inlayHints.variableTypes | false | Whether to display inlay hints for variable types. Accepted values are `true` or `false`. |
 | inlayHints.functionReturnTypes | false |  Whether to display inlay hints for function return types.  Accepted values are `true` or `false`. |
+| inlayHints.pytestParameters | false | Whether to display inlay hints for pytest fixture argument types. Accepted values are `true` or `false`. |
```
Why is this a separate setting instead of just being part of having inlay hints turned on/off for parameters?

luabud replied:
we don't offer type hints for parameters, only variable and function return types 😁

debonte commented on docs/python/testing.md:
```diff
@@ -313,6 +313,21 @@ Below are all the supported commands for testing with the Python extension in VS
+## IntelliSense for pytest fixtures and parameterization
+[Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) offers IntelliSense features that can help you work more efficiently with [pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) and [parameterized tests](https://docs.pytest.org/en/6.2.x/parametrize.html).
+
+As you're typing the parameters for your test function, Pylance will offer you a list of [auto completions](/docs/python/editing.md#autocomplete-and-intellisense) that includes known parameters from arguments defined by `@pytest.mark.parametrize`, as well as existing pytest fixtures defined in your tests file or in `confest.py`. [Code navigation](/docs/python/editing.md#navigation) features such  ...
```
```suggestion
As you're typing the parameters for your test function, Pylance will offer you a list of [auto completions](/docs/python/editing.md#autocomplete-and-intellisense) that includes argument names from `@pytest.mark.parametrize` decorators, as well as existing pytest fixtures defined in your tests file or in `confest.py`. [Code navigation](/docs/python/editing.md#navigation) features such as **Go to Definition** and **Find All References** and [rename symbol refactoring](/docs/editor/refactoring.md#rename-symbol) are also supported.
```

debonte commented on docs/python/testing.md:
```diff
@@ -313,6 +313,21 @@ Below are all the supported commands for testing with the Python extension in VS
+
+![Auto completion suggestion when passing a parameter to a test function in a Python test file. The suggestion is a fixture defined in conftest.py.](/docs/python/images/testing/pytest-fixture-autocomplete.png)
+
+When hovering over a fixture reference or a parameterized argument reference, Pylance will show the inferred type annotation, either based on the return values of the fixture, or based on the inferred types of the arguments passed by the parameterization decorator.
```
```suggestion
When hovering over a fixture reference or a parameterized argument reference, Pylance will show the inferred type annotation, either based on the return values of the fixture, or based on the inferred types of the arguments passed to the parameterization decorator.
```

rchiodo reviewed (changes requested):
I think we should change the screenshot for the parameter type.

luabud commented:
Thank you all very much for you feedback and review suggestions! I think I addressed all comments 😊

@rchiodo thanks for bringing it up, I was completely ignoring the type annotations from the screenshot 🤦‍♀️ I added a different example, let me know if you think that works better
